src/com/oltpbenchmark/benchmarks/wikipedia/WikipediaWorker.java

```diff
@@ -62,58 +60,65 @@ private String generateUserIP() {
 
     @Override
     protected TransactionStatus executeWork(TransactionType nextTransaction) throws UserAbortException, SQLException {
-        WikipediaOperation t = null;
-        
+		Flat z_users = new Flat(this.rng(), 1, this.num_users);
+		Zipf z_pages = new Zipf(this.rng(), 1, this.num_pages, WikipediaConstants.USER_ID_SIGMA);
+
         Class<? extends Procedure> procClass = nextTransaction.getProcedureClass();
         boolean needUser = (procClass.equals(AddWatchList.class) ||
                             procClass.equals(RemoveWatchList.class) ||
-                            procClass.equals(GetPageAuthenticated.class));    
-        while (t == null) {
-            t = this.generator.nextTransaction();
-            if (needUser && t.userId == 0) {
-                t = null;
-            }
-        } // WHILE
-        assert(t != null);
-        if (t.userId != 0) t.userId = this.usersRng.nextInt();
-        
+                            procClass.equals(GetPageAuthenticated.class));
+
+		int userId;
+
+		do {
+			// Check whether this should be an anonymous update
+			if (this.rng().nextInt(100) < WikipediaConstants.ANONYMOUS_PAGE_UPDATE_PROB) {
+				userId = WikipediaConstants.ANONYMOUS_USER_ID;
+			}
+			// Otherwise figure out what user is updating this page
+			else {
+				userId = z_users.nextInt();
+			}
+			// Repeat if we need a user but we generated Anonymous
+		} while (needUser && userId == WikipediaConstants.ANONYMOUS_USER_ID);
+
+		// Figure out what page they're going to update
+		int page_id = z_pages.nextInt();
+
+		Random rand = new Random();
```

Comment:
Again, you don't need to allocate a new `Random` object for every new txn. That is going to be slow. I think you can just use `this.rng`. Double check that it is unique per worker object.

Reply:
> Instances of java.util.Random are threadsafe. However, the concurrent use of the same java.util.Random instance across threads may encounter contention and consequent poor performance. Consider instead using ThreadLocalRandom in multithreaded designs.

- https://docs.oracle.com/javase/8/docs/api/java/util/Random.html

this.rng() refers to this.benchmark.rng() which is a single Random object. By the docs, that should be fine, but we might see worse performance. Perhaps one Random object per thread is still better, or we could refactor to use their suggested ThreadLocalRandom?